### PR TITLE
refactor: mailbox wait_with_progress

### PR DESCRIPTION
**Pull request by timoffex**
Implements `wait_with_progress`, which replaces the old  `MailboxProbe` / `MailboxProgress` mechanism.

Basic usage:

```python
async def display_progress() -> NoReturn:
    # This function never returns. It is cancelled by wait_with_progress
    # on timeout or when the handle gets a result or is abandoned.
    while True:
        update_screen()
        await asyncio.sleep(0.1)

wait_with_progress(
    handle,
    timeout=30,
    progress_after=1,
    display_progress=display_progress,
)
```

Using `asyncio` allows `display_progress` to poll (e.g. RequestPollExit) and update the screen simultaneously---at independent rates!---by starting two tasks (not shown above).

### CODE REVIEW

**Comment by timoffex** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/wandb/wandb/9382?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9382** <a href="https://app.graphite.dev/github/pr/wandb/wandb/9382?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/wandb/wandb/9382?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9380** <a href="https://app.graphite.dev/github/pr/wandb/wandb/9380?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#9379** <a href="https://app.graphite.dev/github/pr/wandb/wandb/9379?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: wandb/sdk/mailbox/wait_with_progress.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+def wait_all_with_progress(
+    handle_list: list[handles.MailboxHandle],
+    *,
+    timeout: float | None,
+    progress_after: float,
+    display_progress: Callable[[], Coroutine[Any, Any, None]],
+) -> list[pb.Result]:
+    """Wait for multiple handles, possibly displaying progress to the user.
+
+    Arguments and exceptions are the same as for `wait_with_progress`.
```
Comment (kptkin) [inline wandb/sdk/mailbox/wait_with_progress.py:58]: nit: given that wait_with_progress calls wait_all_with_progres should the docs be referred to the other way around

File: wandb/sdk/mailbox/wait_with_progress.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+    Returns:
+        A list where the Nth item is the result of the Nth handle.
+    """
+    if not handle_list:
+        return []
+
+    if timeout is not None and timeout <= progress_after:
```
Comment (kptkin) [inline wandb/sdk/mailbox/wait_with_progress.py:66]: could you add more comment about this code path vs the progress with timeout

File: wandb/sdk/mailbox/wait_with_progress.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    """Asynchronously wait for multiple mailbox handles.
+
+    Just like _wait_handles.
+    """
+    results: dict[handles.MailboxHandle, pb.Result] = {}
+
+    async def wait_single(handle: handles.MailboxHandle) -> None:
+        results[handle] = await handle.wait_async(timeout=timeout)
+
+    async with asyncio_compat.open_task_group() as task_group:
```
Comment (kptkin) [inline wandb/sdk/mailbox/wait_with_progress.py:131]: do you think it makes sense to have per task group timeout (i.e. one global on top of each for each task?)

File: wandb/sdk/mailbox/wait_with_progress.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    timeout: float | None,
+) -> list[pb.Result]:
+    """Asynchronously wait for multiple mailbox handles.
+
+    Just like _wait_handles.
+    """
+    results: dict[handles.MailboxHandle, pb.Result] = {}
+
+    async def wait_single(handle: handles.MailboxHandle) -> None:
+        results[handle] = await handle.wait_async(timeout=timeout)
```
Comment (kptkin) [inline wandb/sdk/mailbox/wait_with_progress.py:129]: should we test for edge cases here of overwrite the same handle?

**Review by kptkin — APPROVED** [review]

File: wandb/sdk/mailbox/wait_with_progress.py (hunk 1 above)
Comment (timoffex) [reply wandb/sdk/mailbox/wait_with_progress.py]: done!

File: wandb/sdk/mailbox/wait_with_progress.py (hunk 4 above)
Comment (timoffex) [reply wandb/sdk/mailbox/wait_with_progress.py]: ooh, good catch! fixed this to use handle indices instead and added a test

File: wandb/sdk/mailbox/wait_with_progress.py (hunk 2 above)
Comment (timoffex) [reply wandb/sdk/mailbox/wait_with_progress.py:63]: I considered writing a comment like

> When timeout <= progress_after, we will never show progress.

but this seems obvious. Do you think it would be useful, or is there something else I should elaborate on?

File: wandb/sdk/mailbox/wait_with_progress.py (hunk 3 above)
Comment (timoffex) [reply wandb/sdk/mailbox/wait_with_progress.py:129]: The first timeout from a handle cancels the entire group, so it's not necessary. If I did add a group-level timeout, I would do it by starting a task that times out after `timeout`, which is what this code already does.

File: wandb/sdk/mailbox/wait_with_progress.py (hunk 2 above)
Comment (kptkin) [reply wandb/sdk/mailbox/wait_with_progress.py:63]: no, now that i read it again, i agree this is fine

File: wandb/sdk/mailbox/wait_with_progress.py (hunk 3 above)
Comment (kptkin) [reply wandb/sdk/mailbox/wait_with_progress.py:129]: yeah, that what i understood from the task group semantics, in your previous compat_async PR, but for some reason here i got confused that it won't work properly, ok, so you can ignore it.

**Comment by timoffex** [comment]
### Merge activity

* **Feb 12, 4:11 PM EST**: A user merged this pull request with [Graphite](https://app.graphite.dev/github/pr/wandb/wandb/9382).
